### PR TITLE
feat: choose gpus

### PR DESCRIPTION
**Pull request by keyvank**
#38

### CODE REVIEW

File: src/gpu.rs [hunk 1]
```diff
@@ -778,4 +803,45 @@ mod tests {
...
+
+    #[test]
+    fn test_custom_gpus() {
+        let bus_ids = cl::get_all_bus_ids().unwrap();
+        for bus_id in bus_ids {
+            test_mbatch_hash8s_on_device(
+                cl::futhark_context(cl::GPUSelector::BusId(bus_id)).unwrap(),
+            );
+        }
+    }
```
Comment (porcuquine) [inline src/gpu.rs:846]: This test fails on a machine with two 2080Ti GPUs:
```
---- gpu::tests::test_custom_gpus stdout ----
thread 'gpu::tests::test_custom_gpus' panicked at 'called `Result::unwrap()` on an `Err` value: DeviceNotFound', src/gpu.rs:840:23
```

File: src/cl.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+#[cfg(all(feature = "gpu", not(target_os = "macos")))]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_bus_id_uniqueness() {
+        let devices = get_all_devices().unwrap();
+        let bus_ids = get_all_bus_ids().unwrap();
+        assert_eq!(devices.len(), bus_ids.len());
+    }
```
Comment (porcuquine) [inline src/cl.rs:273]: This test fails on a machine with two 2080Ti GPUs:

```
---- cl::tests::test_bus_id_uniqueness stdout ----
thread 'cl::tests::test_bus_id_uniqueness' panicked at 'called `Result::unwrap()` on an `Err` value: DeviceNotFound', src/cl.rs:270:23
```

File: src/gpu.rs (hunk 1 above)
Comment (keyvank) [reply src/gpu.rs:846]: No idea what's happening in your system. May I check it myself in your system through SSH?

**Comment by porcuquine** [comment]
@keyvank I can now (barely) observe the `test_custom_gpus` test touching both GPUs on my dev machine. However, it's tough to observe and not wholly convincing.

Could you extend the [gbench](https://github.com/filecoin-project/neptune/blob/master/gbench/src/bin/gbench.rs) program to run on all available GPUs? This will require some tweaks to the logging so the logs are still comprehensible (maybe prefix logs with a GPU id), but if it works as expected, we should be able to observe that it completes two runs at once without adversely affecting the timing. If it succeeds but worsens performance, it will still be interesting to see how much and try to understand why.

**Comment by porcuquine** [comment]
@keyvank 

I managed to get past previous problems and have now tested gbench on two GPU (both 2080Ti). We are getting somewhere.

Big problems:
- You can see a wrong result detected and reported.
- The whole thing dies by panic.

Smaller problems:
- The INFO logs don't distinguish the GPU, which makes them difficult or impossible to use to extract timing information.

Request:
- Can you extend so the default is to use just one GPU as previously. This will let me run this without crashing my machine and keep the primary use case simple and unchanged for those who want to benchmark and tune.
- We could also, later, allow for selection of the specific GPU to use, but for now just defaulting to old behavior and only using all GPUs on demand should be enough.

```
➜  gbench git:(choose-gpus) RUST_LOG=info cargo run --release
    Finished release [optimized] target(s) in 0.04s
     Running `target/release/gbench`
[2020-07-01T19:24:16Z INFO  gbench] KiB: 4194304
[2020-07-01T19:24:16Z INFO  gbench] leaves: 134217728
[2020-07-01T19:24:16Z INFO  gbench] max column batch size: 400000
[2020-07-01T19:24:16Z INFO  gbench] max tree batch size: 700000
GPU[Bus-id: 1] --> Run 0
[2020-07-01T19:24:16Z INFO  gbench] Creating ColumnTreeBuilder
GPU[Bus-id: 76] --> Run 0
[2020-07-01T19:24:16Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T19:24:25Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T19:24:25Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T19:24:25Z INFO  gbench] adding column batches
[2020-07-01T19:24:25Z INFO  gbench] start commitment
[2020-07-01T19:24:26Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T19:24:26Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T19:24:26Z INFO  gbench] adding column batches
[2020-07-01T19:24:26Z INFO  gbench] start commitment
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
[2020-07-01T19:26:33Z INFO  gbench] adding final column batch and building tree
..........
[2020-07-01T19:26:38Z INFO  gbench] adding final column batch and building tree
[2020-07-01T19:26:52Z INFO  gbench] end commitment
[2020-07-01T19:26:52Z INFO  gbench] commitment time: 146.998479933s
GPU[Bus-id: 1] --> Run 1
[2020-07-01T19:26:52Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T19:26:54Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T19:26:54Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T19:26:54Z INFO  gbench] adding column batches
[2020-07-01T19:26:54Z INFO  gbench] start commitment
[2020-07-01T19:26:57Z INFO  gbench] end commitment
[2020-07-01T19:26:57Z INFO  gbench] commitment time: 151.147262298s
.......GPU[Bus-id: 76] --> Run 1
[2020-07-01T19:26:57Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T19:26:59Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T19:26:59Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T19:26:59Z INFO  gbench] adding column batches
[2020-07-01T19:26:59Z INFO  gbench] start commitment
....................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
[2020-07-01T19:29:04Z INFO  gbench] adding final column batch and building tree
...................
[2020-07-01T19:29:12Z INFO  gbench] adding final column batch and building tree
[2020-07-01T19:29:22Z INFO  gbench] end commitment
[2020-07-01T19:29:22Z INFO  gbench] commitment time: 147.749961984s
GPU[Bus-id: 1] --> Run 2
[2020-07-01T19:29:23Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T19:29:25Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T19:29:25Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T19:29:25Z INFO  gbench] adding column batches
[2020-07-01T19:29:25Z INFO  gbench] start commitment
[2020-07-01T19:29:31Z INFO  gbench] end commitment
[2020-07-01T19:29:31Z INFO  gbench] commitment time: 151.204301912s
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `Fr(0x0f58900962151ce19bcf1d3aa34e178d6d272a903a5121bcfe57cf90c97ed265)`,
 right: `Fr(0x346955490d6a3ba845e8662be8aca61c43eda5ff990dd8ea138833448abe7df0)`: computed root was not the expected one', src/bin/gbench.rs:85:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
...............................................................................................................................................................................................................................................................................................................................................
[2020-07-01T19:31:36Z INFO  gbench] adding final column batch and building tree
[2020-07-01T19:31:55Z INFO  gbench] end commitment
[2020-07-01T19:31:55Z INFO  gbench] commitment time: 149.840184143s
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/bin/gbench.rs:124:9
```

**Comment by porcuquine** [comment]
@keyvank 

I ran it again. It fails in the same way again, but — interestingly — not with the same wrong value. So something non-deterministic is happening.

```
    Finished release [optimized] target(s) in 0.06s
     Running `target/release/gbench`
[2020-07-01T21:23:52Z INFO  gbench] KiB: 4194304
[2020-07-01T21:23:52Z INFO  gbench] leaves: 134217728
[2020-07-01T21:23:52Z INFO  gbench] max column batch size: 400000
[2020-07-01T21:23:52Z INFO  gbench] max tree batch size: 700000
GPU[Bus-id: 1] --> Run 0
GPU[Bus-id: 76] --> Run 0
[2020-07-01T21:23:52Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:23:52Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:24:00Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:24:00Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:24:00Z INFO  gbench] adding column batches
[2020-07-01T21:24:00Z INFO  gbench] start commitment
[2020-07-01T21:24:01Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:24:01Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:24:01Z INFO  gbench] adding column batches
[2020-07-01T21:24:01Z INFO  gbench] start commitment
.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
[2020-07-01T21:26:09Z INFO  gbench] adding final column batch and building tree
.......
[2020-07-01T21:26:12Z INFO  gbench] adding final column batch and building tree
[2020-07-01T21:26:27Z INFO  gbench] end commitment
[2020-07-01T21:26:27Z INFO  gbench] commitment time: 146.221988552s
GPU[Bus-id: 1] --> Run 1
[2020-07-01T21:26:28Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:26:30Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:26:30Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:26:30Z INFO  gbench] adding column batches
[2020-07-01T21:26:30Z INFO  gbench] start commitment
[2020-07-01T21:26:31Z INFO  gbench] end commitment
[2020-07-01T21:26:31Z INFO  gbench] commitment time: 150.757920347s
...GPU[Bus-id: 76] --> Run 1
[2020-07-01T21:26:31Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:26:33Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:26:33Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:26:33Z INFO  gbench] adding column batches
[2020-07-01T21:26:33Z INFO  gbench] start commitment
..........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
[2020-07-01T21:28:39Z INFO  gbench] adding final column batch and building tree
.................
[2020-07-01T21:28:45Z INFO  gbench] adding final column batch and building tree
[2020-07-01T21:28:57Z INFO  gbench] end commitment
[2020-07-01T21:28:57Z INFO  gbench] commitment time: 147.032292118s
GPU[Bus-id: 1] --> Run 2
[2020-07-01T21:28:57Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:28:59Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:28:59Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:28:59Z INFO  gbench] adding column batches
[2020-07-01T21:28:59Z INFO  gbench] start commitment
[2020-07-01T21:29:04Z INFO  gbench] end commitment
[2020-07-01T21:29:04Z INFO  gbench] commitment time: 150.822873474s
.............GPU[Bus-id: 76] --> Run 2
[2020-07-01T21:29:05Z INFO  gbench] Creating ColumnTreeBuilder
[2020-07-01T21:29:07Z INFO  gbench] ColumnTreeBuilder created
[2020-07-01T21:29:07Z INFO  gbench] Using effective batch size 400000 to build columns
[2020-07-01T21:29:07Z INFO  gbench] adding column batches
[2020-07-01T21:29:07Z INFO  gbench] start commitment
.......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
[2020-07-01T21:31:08Z INFO  gbench] adding final column batch and building tree
..........................
[2020-07-01T21:31:19Z INFO  gbench] adding final column batch and building tree
[2020-07-01T21:31:27Z INFO  gbench] end commitment
[2020-07-01T21:31:27Z INFO  gbench] commitment time: 147.560413122s
[2020-07-01T21:31:38Z INFO  gbench] end commitment
[2020-07-01T21:31:38Z INFO  gbench] commitment time: 151.196229368s
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `Fr(0x0f58900962151ce19bcf1d3aa34e178d6d272a903a5121bcfe57cf90c97ed265)`,
 right: `Fr(0x0f49eaf847841c7a00a11e965ce85ac841ea7ba6d3b44186c8c865ea231d7037)`: computed root was not the expected one', src/bin/gbench.rs:85:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any', src/bin/gbench.rs:124:9
```

**Comment by porcuquine** [comment]
It was eventually determined that the problem above was due to a faulty GPU fan leading to a specific card overheating and therefore producing non-deterministic failures occasionally (but often enough to be easily reproducible with time). With a new GPU replacing the faulty one, we can no longer reproduce these failures.

File: src/cl.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+pub fn futhark_context(selector: GPUSelector) -> ClResult<Arc<Mutex<FutharkContext>>> {
+    let mut map = FUTHARK_CONTEXT_MAP.lock().unwrap();
+    let bus_id = selector.get_bus_id()?;
+    if !map.contains_key(&bus_id) {
+        let device = get_device_by_bus_id(bus_id)?;
+        let context = create_futhark_context(device)?;
+        map.insert(bus_id, Arc::new(Mutex::new(context)));
+    }
+    Ok(Arc::clone(&map[&bus_id]))
```
Comment (porcuquine) [inline src/cl.rs:282]: I think you can do this with `Entry::or_insert_with()`.

File: src/cl.rs [hunk 4]
```diff
@@ -274,12 +274,11 @@ pub fn get_all_bus_ids() -> ClResult<Vec<u32>> {
 pub fn futhark_context(selector: GPUSelector) -> ClResult<Arc<Mutex<FutharkContext>>> {
     let mut map = FUTHARK_CONTEXT_MAP.lock().unwrap();
     let bus_id = selector.get_bus_id()?;
-    if !map.contains_key(&bus_id) {
+    Ok(Arc::clone(map.entry(bus_id).or_insert({
```
Comment (porcuquine) [inline src/cl.rs:277]: No, you need `or_insert_with`. This will always create a new Futhark context to pass as its argument, which we do *not* want.

**Pull request merged**